File: app/src/main/java/org/stepic/droid/ui/dialogs/VideoQualityDialog.java
package org.stepic.droid.ui.dialogs;

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.AlertDialog;

import org.jetbrains.annotations.NotNull;
import org.stepic.droid.R;
import org.stepic.droid.analytic.Analytic;
import org.stepic.droid.base.MainApplication;
import org.stepic.droid.core.IShell;
import org.stepic.droid.preferences.UserPreferences;
import org.stepic.droid.store.operations.DatabaseFacade;

import java.util.HashMap;
import java.util.Map;

import javax.inject.Inject;

public class VideoQualityDialog extends DialogFragment {


    @Inject
    IShell shell;

    @Inject
    DatabaseFacade databaseFacade;

    @Inject
    UserPreferences userPreferences;

    @Inject
    Analytic analytic;

    private Map<String, Integer> mQualityToPositionMap = null;
    private Map<Integer, String> mPositionToQualityMap = null;

    static {
    }


    @NotNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        MainApplication.component().inject(this);
        if (mQualityToPositionMap == null || mPositionToQualityMap == null) {
            initMaps();
        }

        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setTitle(R.string.video_quality)
                .setNegativeButton(R.string.cancel, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        analytic.reportEvent(Analytic.Interaction.CANCEL_VIDEO_QUALITY);
                    }
                })
                .setSingleChoiceItems(R.array.video_quality,
                        mQualityToPositionMap.get(userPreferences.getQualityVideo()),
                        new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, final int which) {
                                analytic.reportEventWithIdName(Analytic.Preferences.VIDEO_QUALITY, which + "", mPositionToQualityMap.get(which));
                                AsyncTask<Void, Void, Void> task = new AsyncTask<Void, Void, Void>() {
                                    @Override
                                    protected Void doInBackground(Void... params) {
                                        userPreferences.storeQualityVideo(mPositionToQualityMap.get(which));
                                        return null;
                                    }
                                };
                                task.execute();
                                dialog.dismiss();
                            }
                        });

        return builder.create();
    }

    private void initMaps() {

        if (mQualityToPositionMap == null) {
            mQualityToPositionMap = new HashMap<>();
            mQualityToPositionMap.put("270", 0);
            mQualityToPositionMap.put("360", 1);
            mQualityToPositionMap.put("720", 2);
            mQualityToPositionMap.put("1080", 3);
        }

        if (mPositionToQualityMap == null) {
            mPositionToQualityMap = new HashMap<>();
            mPositionToQualityMap.put(0, "270");
            mPositionToQualityMap.put(1, "360");
            mPositionToQualityMap.put(2, "720");
            mPositionToQualityMap.put(3, "1080");
        }
    }
}

File: app/src/main/java/org/stepic/droid/ui/dialogs/VideoQualityDialog.kt
package org.stepic.droid.ui.dialogs

import android.app.Dialog
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import android.widget.ArrayAdapter
import org.stepic.droid.R
import org.stepic.droid.analytic.Analytic
import org.stepic.droid.base.App
import org.stepic.droid.preferences.UserPreferences
import ru.nobird.android.view.base.ui.extension.argument
import java.util.concurrent.ThreadPoolExecutor
import javax.inject.Inject

class VideoQualityDialog : VideoQualityDialogBase() {
    companion object {
        fun newInstance(forPlaying: Boolean) =
            VideoQualityDialog().also {
                it.forPlaying = forPlaying
            }
    }

    @Inject
    lateinit var userPreferences: UserPreferences

    @Inject
    lateinit var analytic: Analytic

    @Inject
    lateinit var threadPoolExecutor: ThreadPoolExecutor

    private var forPlaying by argument<Boolean>()

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        init()

        val qualityValue = if (forPlaying) {
            userPreferences.qualityVideoForPlaying
        } else {
            userPreferences.qualityVideo
        }

        val chosenOptionPosition = qualityToPositionMap[qualityValue]!!

        val adapter = ArrayAdapter<String>(requireContext(), R.layout.simple_list_item_single_choice, resources.getStringArray(R.array.video_quality))

        val builder = AlertDialog.Builder(requireContext())
        builder
            .setTitle(
                if (forPlaying) {
                    R.string.video_quality_playing
                }
                else {
                    R.string.video_quality
                }
            )
            .setSingleChoiceItems(adapter, chosenOptionPosition) { dialog, which ->
                val qualityString = positionToQualityMap[which]
                analytic.reportEventWithIdName(Analytic.Preferences.VIDEO_QUALITY, which.toString(), qualityString)

                threadPoolExecutor.execute {
                    if (forPlaying) {
                        userPreferences.saveVideoQualityForPlaying(qualityString)
                    } else {
                        userPreferences.storeQualityVideo(qualityString)
                    }
                }
                dialog.dismiss()
            }
            .setNegativeButton(R.string.cancel) { _, _ ->
                analytic.reportEvent(Analytic.Interaction.CANCEL_VIDEO_QUALITY)
            }

        return builder.create()
    }

    override fun injectDependencies() {
        App.component().inject(this)
    }
}
